cpp/src/fil/infer.cu

```diff
@@ -480,6 +487,115 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES> {
+                                               void* finalize_workspace,
+                                               float* vector_leaf)
+    : num_classes(params.num_classes),
+      num_threads_per_class(max(1, blockDim.x / params.num_classes)),
+      vector_leaf(vector_leaf),
+      tmp_storage(finalize_workspace) {
+    // Assign workspace
+    char* ptr = (char*)accumulate_workspace;
+    per_class_margin = (vec<NITEMS, float>*)ptr;
+    ptr += sizeof(vec<NITEMS, float>) * num_classes * num_threads_per_class;
+    vector_leaf_indices = (vec<NITEMS, int>*)ptr;
+    ptr += sizeof(vec<NITEMS, int>) * blockDim.x;
+    thread_num_rows = (int*)ptr;
+
+    // Initialise shared memory
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      per_class_margin[i] = vec<NITEMS, float>();
+    }
+    vector_leaf_indices[threadIdx.x] = vec<NITEMS, int>();
+    thread_num_rows[threadIdx.x] = 0;
+    // __syncthreads() is called in infer_k
+  }
+
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, int> single_tree_prediction, int tree, int num_rows) {
+    // Perform a transpose in shared memory
+    // Assign each thread to a class, so they can accumulate without atomics
+    __syncthreads();
+    // Write indices to shared memory
+    vector_leaf_indices[threadIdx.x] = single_tree_prediction;
+    thread_num_rows[threadIdx.x] = num_rows;
+    __syncthreads();
+    // i here refers to each element of the matrix per_class_margin
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      // if num_threads_per_class == 1, then c == i
+      int c = i % num_classes;
+      int row = i / num_classes;
+      // iterate over original thread inputs with stride num_threads_per_class
+      // j is the original thread input
+      // we have num_classes threads for each j
+      for (int j = row; j < blockDim.x; j += num_threads_per_class) {
+        for (int item = 0; item < thread_num_rows[j]; ++item) {
+          float pred =
+            vector_leaf[vector_leaf_indices[j][item] * num_classes + c];
+          per_class_margin[i][item] += pred;
+        }
+      }
+    }
+  }
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int num_outputs, output_t transform,
+                                           int num_trees,
+                                           int log2_threads_per_tree) {
+    __syncthreads();
+    if (num_classes < blockDim.x) {
+      // Efficient implementation for small number of classes
+      auto acc = multi_sum<6>(per_class_margin, num_classes,
+                              max(1, blockDim.x / num_classes));
+      if (threadIdx.x < num_classes) per_class_margin[threadIdx.x] = acc;
+    } else {
+      // For larger numbers of classes
+      for (int c = threadIdx.x; c < num_classes; c += blockDim.x) {
+        for (int i = threadIdx.x + num_classes;
```

Comment:
isn't this an empty loop? Since `num_threads_per_class == 1` here. There shouldn't be anything to reduce if there weren't enough threads to cover the classes even once.